File: leetcode/solutions/0219.ContainsDuplicateII/containsNearbyDuplicate.go
package solutions

func containsNearbyDuplicate(nums []int, k int) bool {
	indexMap := make(map[int]int, len(nums))

	for i, num := range nums {
		if index, found := indexMap[num]; found && i-index <= k {
			return true
		}

		indexMap[num] = i
	}

	return false
}

File: leetcode/solutions/0219.ContainsDuplicateII/containsNearbyDuplicate.ts
export default function containsNearbyDuplicate(
  nums: number[],
  k: number
): boolean {
  // map form: [nums[i], i]
  const lastIndices = new Map<number, number>();

  for (const [i, num] of nums.entries()) {
    if (i - (lastIndices.get(num) ?? -Infinity) <= k) {
      return true;
    }

    lastIndices.set(num, i);
  }

  return false;
}
